layabout.py

```diff
@@ -240,51 +256,69 @@ def run(self, *, token: str = None, interval: float = 0.5,
+        self._ensure_slack(
+            connector=connector,
+            retries=retries,
+            backoff=backoff
+        )
+        assert self._slack is not None
 
         while True:
-            try:
-                # Fetch new RTM events from the API.
-                events = self._slack.rtm_read()
-
-            # This is necessary to handle an error caused by a bug in Slack's
-            # Python client. For more information see
-            # https://github.com/slackhq/python-slackclient/issues/127
-            #
-            # TODO: The TimeoutError could be more elegantly resolved by making
-            # a PR to the websocket-client library and letting them coerce that
-            # exception to a WebSocketTimeoutException.
-            except (WebSocketConnectionClosedException, TimeoutError):
-                log.debug('Lost connection to the Slack API, attempting to '
-                          'reconnect')
-                # Attempt to reconnect with our existing SlackClient and token.
-                if self._reconnect(retries=retries, backoff=backoff):
-                    continue
-
-                raise FailedConnection('Failed to reconnect to the Slack API')
+            events = self._slack.fetch_events()
 
             if not until(events):
                 log.debug('Exiting event loop')
                 break
 
             # Handle events!
             for event in events:
-                type_ = event.get('type')
+                type_ = event.get('type', '')
                 for handler in self._handlers[type_] + self._handlers['*']:
                     fn, kwargs = handler
-                    fn(self._slack, event, **kwargs)
+                    fn(self._slack.inner, event, **kwargs)
 
             # Maybe don't pester the Slack API too much.
             time.sleep(interval)
 
 
-def _forever(events: List[dict]) -> bool:  # pragma: no cover
+@singledispatch
+def _create_slack(connector: Any):
+    """ Default connector. Raises an error with unsupported connectors. """
+    raise TypeError(f"Invalid connector: {type(connector)}")
+
+
+@_create_slack.register(EnvVar)
+def _create_slack_with_env_var(env_var: EnvVar) -> SlackClient:
+    """ Create a SlackClient with a token from an env var. """
+    token = os.getenv(env_var)
+    if not token:
+        raise MissingSlackToken("Could not acquire token from "
+                                f"{env_var}")
+
+    return SlackClient(token=token)
+
+
+@_create_slack.register(str)
+def _create_slack_with_token(token: str) -> SlackClient:
+    """ Create a SlackClient with a provided token. """
+    if token == '':
+        raise MissingSlackToken("The empty string is an invalid Slack API "
+                                "token")
+
+    return SlackClient(token=token)
+
+
+@_create_slack.register(SlackClient)
+def _create_slack_with_slack_client(slack: SlackClient) -> SlackClient:
+    """ Use an existing SlackClient if we don't already have one. """
```

Comment:
This comment is wrong now.  Just, use the existing slack client if the user passes us one